File: apps/mo1-ios/src/api-client.ts
// api-client — MO3 `/m/v1/*` client (design §2-2 ApiClient module). Owns the
// three cross-cutting behaviours the Swift ApiClient must match exactly:
//   1. Bearer attach from the Keychain-backed TokenStore.
//   2. 401 -> single silent refresh -> retry ONCE (design §6, test §7 "1回性").
//   3. ErrorResponse -> semi-open DubClientError + retryable backoff (max 2).
// Contract types come from @dub/types `mobile`/`task`/`event`/`notification`
// namespaces (MO3 正本); this layer never re-defines them.
import {
  common,
  type auth,
  type chat,
  type event,
  type gantt,
  type mobile,
  type notification,
  type task,
} from "@dub/types";
import { CommonErrorCodes, isErrorResponse } from "@dub/errors";
import type { Transport, HttpMethod } from "./transport.js";
import type { TokenStore, StoredSession } from "./token-store.js";
import { fromResponseBody, offlineError, type DubClientError } from "./errors.js";

/**
 * Mobile auth envelope. The MO3 `mobile` namespace has not yet frozen the
 * exchange/refresh RESPONSE shape (only auth.MobileExchangeRequest exists), so
 * this is a client-local view over the frozen auth.SessionInfo — replace with
 * the generated `mobile.MobileAuthSession` once MO3 publishes it (see notes).
 */
export interface MobileAuthSession {
  token: string;
  session: auth.SessionInfo;
}

export interface ApiClientConfig {
  /** origin of m-api.developershub.jp, e.g. "https://m-api.developershub.jp". */
  baseUrl: string;
  transport: Transport;
  tokenStore: TokenStore;
  /** max retryable-error retries (default 2, design §6). */
  maxRetries?: number;
  /** injectable delay so tests don't wait real time. */
  sleep?: (ms: number) => Promise<void>;
  /** called after refresh fails / retry still 401: UI routes to S1 login. */
  onSessionExpired?: () => void;
}

interface RequestOptions {
  query?: Record<string, string | number | boolean | undefined>;
  body?: unknown;
  /** skip the Bearer header + 401-refresh dance (auth exchange/refresh/login). */
  anonymous?: boolean;
}

const defaultSleep = (ms: number): Promise<void> => new Promise((r) => setTimeout(r, ms));

export class MobileApiClient {
  readonly #cfg: Required<Omit<ApiClientConfig, "onSessionExpired">> & Pick<ApiClientConfig, "onSessionExpired">;
  /** de-dupes concurrent 401s into a single in-flight refresh. */
  #refreshInFlight: Promise<boolean> | null = null;

  constructor(cfg: ApiClientConfig) {
    this.#cfg = {
      maxRetries: 2,
      sleep: defaultSleep,
      ...cfg,
    };
  }

  // ---- typed endpoint wrappers (design §2-1) --------------------------------

  exchange(req: auth.MobileExchangeRequest): Promise<MobileAuthSession> {
    return this.#request<MobileAuthSession>("POST", "/auth/exchange", { body: req, anonymous: true });
  }

  logout(): Promise<void> {
    const session = this.#cfg.tokenStore.read();
    const p = this.#request<null>("POST", "/auth/logout", { body: {} }).then(() => undefined);
    // logout always clears local state regardless of server outcome.
    return p.catch(() => undefined).finally(() => {
      if (session) this.#cfg.tokenStore.clear();
    });
  }

  getHome(): Promise<mobile.MobileHomeResponse> {
    return this.#request<mobile.MobileHomeResponse>("GET", "/bff/home");
  }

  getEventOverview(eventId: string): Promise<mobile.MobileEventOverviewResponse> {
    return this.#request<mobile.MobileEventOverviewResponse>("GET", `/events/${encodeURIComponent(eventId)}`);
  }

  listTasks(query: task.ListTasksQuery = {}): Promise<task.ListTasksResponse> {
    return this.#request<task.ListTasksResponse>("GET", "/tasks", {
      query: {
        cursor: query.cursor,
        limit: query.limit,
        eventId: query.eventId,
        assigneeId: query.assigneeId,
      },
    });
  }

  patchTask(id: string, req: task.UpdateTaskRequest): Promise<task.Task> {
    return this.#request<task.Task>("PATCH", `/tasks/${encodeURIComponent(id)}`, { body: req });
  }

  getInbox(query: notification.ListInboxQuery = {}): Promise<notification.ListInboxResponse> {
    return this.#request<notification.ListInboxResponse>("GET", "/inbox", {
      query: { cursor: query.cursor, limit: query.limit },
    });
  }

  /** S6 gantt chart read: rows + FS dependency lines for one event. */
  getGantt(query: gantt.GetGanttQuery): Promise<gantt.GanttChartDTO> {
    return this.#request<gantt.GanttChartDTO>("GET", "/gantt", {
      query: { eventId: query.eventId },
    });
  }

  /** S8 chat: list the channels the caller can see. */
  listChatChannels(query: common.CursorQuery = {}): Promise<common.Paginated<chat.ChatChannel>> {
    return this.#request<common.Paginated<chat.ChatChannel>>("GET", "/chat/channels", {
      query: { cursor: query.cursor, limit: query.limit },
    });
  }

  /** S8 chat: page a channel's message history (oldest resolved via cursor). */
  listChatMessages(
    channelId: string,
    query: common.CursorQuery = {},
  ): Promise<common.Paginated<chat.ChatMessage>> {
    return this.#request<common.Paginated<chat.ChatMessage>>(
      "GET",
      `/chat/channels/${encodeURIComponent(channelId)}/messages`,
      { query: { cursor: query.cursor, limit: query.limit } },
    );
  }

  /** S8 chat: short-lived ticket for the DO-direct (gateway-bypassing) WS. */
  getChatWsTicket(channelId: string): Promise<chat.WsTicketResponse> {
    return this.#request<chat.WsTicketResponse>(
      "GET",
      `/chat/channels/${encodeURIComponent(channelId)}/ws-ticket`,
    );
  }

  registerDevice(req: mobile.RegisterDeviceRequest): Promise<mobile.RegisterDeviceResponse> {
    return this.#request<mobile.RegisterDeviceResponse>("POST", "/devices", { body: req });
  }

  deleteDevice(deviceId: string): Promise<void> {
    return this.#request<null>("DELETE", `/devices/${encodeURIComponent(deviceId)}`).then(() => undefined);
  }

  // ---- core ----------------------------------------------------------------

  async #request<T>(method: HttpMethod, path: string, opts: RequestOptions = {}): Promise<T> {
    const url = this.#buildUrl(path, opts.query);
    let attempt = 0;
    let refreshedOnce = false;

    for (;;) {
      let res;
      try {
        res = await this.#cfg.transport({
          method,
          url,
          headers: this.#headers(opts.anonymous ?? false),
          ...(opts.body !== undefined ? { body: opts.body } : {}),
        });
      } catch (cause) {
        // transport failure = offline; retry within budget then surface.
        if (attempt < this.#cfg.maxRetries) {
          await this.#backoff(attempt++);
          continue;
        }
        throw offlineError(cause);
      }

      if (res.status >= 200 && res.status < 300) {
        return res.body as T;
      }

      const err = fromResponseBody(res.status, res.body);

      // 401 -> one silent refresh -> retry once (only on authenticated calls).
      if (err.kind === "reauth" && !opts.anonymous) {
        if (!refreshedOnce) {
          refreshedOnce = true;
          const ok = await this.#ensureRefreshed();
          if (ok) continue; // retry with the new token
        }
        // refresh failed, or the retry still 401 -> session is unusable.
        this.#cfg.tokenStore.clear();
        this.#cfg.onSessionExpired?.();
        throw err;
      }

      // retryable upstream / rate-limit -> exponential backoff (Retry-After wins).
      if (err.retryable && attempt < this.#cfg.maxRetries) {
        await this.#backoff(attempt++, err.retryAfterSec);
        continue;
      }

      throw err;
    }
  }

  /** Single-flight refresh: many concurrent 401s share one refresh call. */
  #ensureRefreshed(): Promise<boolean> {
    if (this.#refreshInFlight) return this.#refreshInFlight;
    const p = this.#doRefresh().finally(() => {
      this.#refreshInFlight = null;
    });
    this.#refreshInFlight = p;
    return p;
  }

  async #doRefresh(): Promise<boolean> {
    const current = this.#cfg.tokenStore.read();
    if (!current) return false;
    let res;
    try {
      res = await this.#cfg.transport({
        method: "POST",
        url: this.#buildUrl("/auth/refresh"),
        // Bearer path (theme8): current token is the refresh credential.
        headers: { ...this.#jsonHeaders(), authorization: `Bearer ${current.token}` },
        body: {} satisfies auth.AuthRefreshRequest,
      });
    } catch {
      return false;
    }
    if (res.status < 200 || res.status >= 300) return false;
    if (!isMobileAuthSession(res.body)) return false;
    const next: StoredSession = {
      token: res.body.token,
      sessionExpiresAt: res.body.session.sessionExpiresAt,
    };
    this.#cfg.tokenStore.write(next);
    return true;
  }

  #buildUrl(path: string, query?: RequestOptions["query"]): string {
    const base = this.#cfg.baseUrl.replace(/\/+$/, "");
    let url = `${base}${common.MOBILE_API_PREFIX}${path}`;
    if (query) {
      const qs = new URLSearchParams();
      for (const [k, v] of Object.entries(query)) {
        if (v !== undefined) qs.set(k, String(v));
      }
      const s = qs.toString();
      if (s) url += `?${s}`;
    }
    return url;
  }

  #jsonHeaders(): Record<string, string> {
    return { "content-type": "application/json", accept: "application/json" };
  }

  #headers(anonymous: boolean): Record<string, string> {
    const h = this.#jsonHeaders();
    if (!anonymous) {
      const s = this.#cfg.tokenStore.read();
      if (s) h.authorization = `Bearer ${s.token}`;
    }
    return h;
  }

  async #backoff(attempt: number, retryAfterSec?: number): Promise<void> {
    const ms = retryAfterSec !== undefined ? retryAfterSec * 1000 : 2 ** attempt * 200;
    await this.#cfg.sleep(ms);
  }
}

function isMobileAuthSession(v: unknown): v is MobileAuthSession {
  if (v === null || typeof v !== "object") return false;
  const o = v as { token?: unknown; session?: unknown };
  if (typeof o.token !== "string") return false;
  if (o.session === null || typeof o.session !== "object") return false;
  const s = o.session as { sessionExpiresAt?: unknown };
  return typeof s.sessionExpiresAt === "number";
}

/** Re-export for callers that want to narrow thrown errors. */
export type { DubClientError };
export { isErrorResponse, CommonErrorCodes };

File: apps/mo1-ios/ios/Sources/Mo1Core/Networking/ApiClient.swift
// ApiClient — MO3 `/m/v1/*` client (design §2-2). Owns the three cross-cutting
// behaviours that must match the TS reference (api-client.ts) exactly:
//   1. Bearer attach from the Keychain-backed TokenStore.
//   2. 401 -> single silent refresh -> retry ONCE (design §6, "1回性").
//   3. ErrorResponse -> semi-open DubClientError + retryable backoff (max 2).
import Foundation

public protocol MobileApi: Sendable {
    func exchange(_ req: MobileExchangeRequest) async throws -> MobileAuthSession
    func logout() async
    func getHome() async throws -> MobileHomeResponse
    func getEventOverview(_ eventId: String) async throws -> MobileEventOverviewResponse
    func listTasks(_ query: ListTasksQuery) async throws -> ListTasksResponse
    func patchTask(_ id: String, _ req: UpdateTaskRequest) async throws -> DubTask
    func getInbox(_ query: ListInboxQuery) async throws -> ListInboxResponse
    func registerDevice(_ req: RegisterDeviceRequest) async throws -> RegisterDeviceResponse
    func deleteDevice(_ deviceId: String) async throws
    func getGantt(_ query: GetGanttQuery) async throws -> GanttChartDTO
    func listChatChannels(_ query: CursorQuery) async throws -> ListChatChannelsResponse
    func listChatMessages(_ channelId: String, _ query: CursorQuery) async throws -> ListChatMessagesResponse
    func getChatWsTicket(_ channelId: String) async throws -> WsTicketResponse
}

public struct ApiClientConfig {
    /// origin of m-api.developershub.jp, e.g. "https://m-api.developershub.jp".
    public var baseURL: String
    public var transport: Transport
    public var tokenStore: TokenStore
    /// max retryable-error retries (default 2, design §6).
    public var maxRetries: Int
    /// injectable delay (ms) so tests don't wait real time.
    public var sleep: @Sendable (UInt64) async -> Void
    /// called after refresh fails / retry still 401: UI routes to S1 login.
    public var onSessionExpired: (@Sendable () -> Void)?

    public init(
        baseURL: String,
        transport: Transport,
        tokenStore: TokenStore,
        maxRetries: Int = 2,
        sleep: @escaping @Sendable (UInt64) async -> Void = { ms in try? await Task.sleep(nanoseconds: ms * 1_000_000) },
        onSessionExpired: (@Sendable () -> Void)? = nil
    ) {
        self.baseURL = baseURL
        self.transport = transport
        self.tokenStore = tokenStore
        self.maxRetries = maxRetries
        self.sleep = sleep
        self.onSessionExpired = onSessionExpired
    }
}

public final class MobileApiClient: MobileApi, @unchecked Sendable {
    private let cfg: ApiClientConfig
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()
    /// de-dupes concurrent 401s into a single in-flight refresh.
    private var refreshInFlight: Task<Bool, Never>?
    private let lock = NSLock()

    public init(_ cfg: ApiClientConfig) { self.cfg = cfg }

    // ---- typed endpoint wrappers (design §2-1) ------------------------------

    public func exchange(_ req: MobileExchangeRequest) async throws -> MobileAuthSession {
        try await request(.POST, "/auth/exchange", bodyData: try encoder.encode(req), anonymous: true)
    }

    public func logout() async {
        let session = cfg.tokenStore.read()
        // logout always clears local state regardless of server outcome.
        try? await requestVoid(.POST, "/auth/logout", bodyData: Data("{}".utf8))
        if session != nil { cfg.tokenStore.clear() }
    }

    public func getHome() async throws -> MobileHomeResponse {
        try await request(.GET, "/bff/home")
    }

    public func getEventOverview(_ eventId: String) async throws -> MobileEventOverviewResponse {
        try await request(.GET, "/events/\(Self.encodeComponent(eventId))")
    }

    public func listTasks(_ query: ListTasksQuery = ListTasksQuery()) async throws -> ListTasksResponse {
        try await request(.GET, "/tasks", query: [
            "cursor": query.cursor,
            "limit": query.limit.map(String.init),
            "eventId": query.eventId,
            "assigneeId": query.assigneeId,
        ])
    }

    public func patchTask(_ id: String, _ req: UpdateTaskRequest) async throws -> DubTask {
        try await request(.PATCH, "/tasks/\(Self.encodeComponent(id))", bodyData: try encoder.encode(req))
    }

    public func getInbox(_ query: ListInboxQuery = ListInboxQuery()) async throws -> ListInboxResponse {
        try await request(.GET, "/inbox", query: [
            "cursor": query.cursor,
            "limit": query.limit.map(String.init),
        ])
    }

    public func registerDevice(_ req: RegisterDeviceRequest) async throws -> RegisterDeviceResponse {
        try await request(.POST, "/devices", bodyData: try encoder.encode(req))
    }

    public func deleteDevice(_ deviceId: String) async throws {
        try await requestVoid(.DELETE, "/devices/\(Self.encodeComponent(deviceId))")
    }

    // ---- S6 gantt / S8 chat reads (design §2-1) -----------------------------

    /// S6 gantt chart read: rows + FS dependency lines for one event.
    public func getGantt(_ query: GetGanttQuery) async throws -> GanttChartDTO {
        try await request(.GET, "/gantt", query: ["eventId": query.eventId])
    }

    /// S8 chat: list the channels the caller can see.
    public func listChatChannels(_ query: CursorQuery = CursorQuery()) async throws -> ListChatChannelsResponse {
        try await request(.GET, "/chat/channels", query: [
            "cursor": query.cursor,
            "limit": query.limit.map(String.init),
        ])
    }

    /// S8 chat: page a channel's message history (oldest resolved via cursor).
    public func listChatMessages(_ channelId: String, _ query: CursorQuery = CursorQuery()) async throws -> ListChatMessagesResponse {
        try await request(.GET, "/chat/channels/\(Self.encodeComponent(channelId))/messages", query: [
            "cursor": query.cursor,
            "limit": query.limit.map(String.init),
        ])
    }

    /// S8 chat: short-lived ticket for the DO-direct (gateway-bypassing) WS.
    public func getChatWsTicket(_ channelId: String) async throws -> WsTicketResponse {
        try await request(.GET, "/chat/channels/\(Self.encodeComponent(channelId))/ws-ticket")
    }

    // ---- core ---------------------------------------------------------------

    private func request<T: Decodable>(
        _ method: HttpMethod, _ path: String,
        query: [String: String?] = [:], bodyData: Data? = nil, anonymous: Bool = false
    ) async throws -> T {
        let data = try await perform(method, path, query: query, bodyData: bodyData, anonymous: anonymous)
        return try decoder.decode(T.self, from: data)
    }

    private func requestVoid(
        _ method: HttpMethod, _ path: String,
        query: [String: String?] = [:], bodyData: Data? = nil, anonymous: Bool = false
    ) async throws {
        _ = try await perform(method, path, query: query, bodyData: bodyData, anonymous: anonymous)
    }

    private func perform(
        _ method: HttpMethod, _ path: String,
        query: [String: String?], bodyData: Data?, anonymous: Bool
    ) async throws -> Data {
        let url = buildURL(path, query)
        var attempt = 0
        var refreshedOnce = false

        while true {
            let res: TransportResponse
            do {
                res = try await cfg.transport.send(TransportRequest(
                    method: method, url: url, headers: headers(anonymous: anonymous), body: bodyData
                ))
            } catch {
                // transport failure = offline; retry within budget then surface.
                if attempt < cfg.maxRetries {
                    await backoff(attempt); attempt += 1; continue
                }
                throw ErrorMapper.offlineError(error)
            }

            if res.status >= 200 && res.status < 300 {
                return res.body
            }

            let err = ErrorMapper.fromResponseBody(status: res.status, body: res.body)

            // 401 -> one silent refresh -> retry once (authenticated calls only).
            if err.kind == .reauth && !anonymous {
                if !refreshedOnce {
                    refreshedOnce = true
                    if await ensureRefreshed() { continue } // retry with new token
                }
                // refresh failed, or retry still 401 -> session is unusable.
                cfg.tokenStore.clear()
                cfg.onSessionExpired?()
                throw err
            }

            // retryable upstream / rate-limit -> backoff (Retry-After wins).
            if err.retryable && attempt < cfg.maxRetries {
                await backoff(attempt, retryAfterSec: err.retryAfterSec); attempt += 1; continue
            }

            throw err
        }
    }

    /// Single-flight refresh: many concurrent 401s share one refresh call.
    private func ensureRefreshed() async -> Bool {
        let task = inFlightRefreshTask()
        let result = await task.value
        clearRefreshTask()
        return result
    }

    // Lock/unlock kept in synchronous helpers so they never straddle an await.
    private func inFlightRefreshTask() -> Task<Bool, Never> {
        lock.lock(); defer { lock.unlock() }
        if let existing = refreshInFlight { return existing }
        let task = Task { await self.doRefresh() }
        refreshInFlight = task
        return task
    }

    private func clearRefreshTask() {
        lock.lock(); defer { lock.unlock() }
        refreshInFlight = nil
    }

    private func doRefresh() async -> Bool {
        guard let current = cfg.tokenStore.read() else { return false }
        let res: TransportResponse
        do {
            var h = jsonHeaders()
            h["authorization"] = "Bearer \(current.token)" // theme8: current token is the refresh credential
            res = try await cfg.transport.send(TransportRequest(
                method: .POST, url: buildURL("/auth/refresh", [:]), headers: h,
                body: (try? encoder.encode(AuthRefreshRequest())) ?? Data("{}".utf8)
            ))
        } catch {
            return false
        }
        guard res.status >= 200 && res.status < 300,
              let session = try? decoder.decode(MobileAuthSession.self, from: res.body)
        else { return false }
        cfg.tokenStore.write(StoredSession(token: session.token, sessionExpiresAt: session.session.sessionExpiresAt))
        return true
    }

    private func buildURL(_ path: String, _ query: [String: String?]) -> String {
        var base = cfg.baseURL
        while base.hasSuffix("/") { base.removeLast() }
        var url = "\(base)\(MOBILE_API_PREFIX)\(path)"
        let pairs = query.compactMap { key, value -> String? in
            guard let value else { return nil }
            let ek = key.addingPercentEncoding(withAllowedCharacters: Self.queryAllowed) ?? key
            let ev = value.addingPercentEncoding(withAllowedCharacters: Self.queryAllowed) ?? value
            return "\(ek)=\(ev)"
        }.sorted()
        if !pairs.isEmpty { url += "?\(pairs.joined(separator: "&"))" }
        return url
    }

    private func jsonHeaders() -> [String: String] {
        ["content-type": "application/json", "accept": "application/json"]
    }

    private func headers(anonymous: Bool) -> [String: String] {
        var h = jsonHeaders()
        if !anonymous, let s = cfg.tokenStore.read() {
            h["authorization"] = "Bearer \(s.token)"
        }
        return h
    }

    private func backoff(_ attempt: Int, retryAfterSec: Double? = nil) async {
        let ms: UInt64
        if let retryAfterSec {
            ms = UInt64(retryAfterSec * 1000)
        } else {
            ms = UInt64(pow(2.0, Double(attempt)) * 200)
        }
        await cfg.sleep(ms)
    }

    private static let queryAllowed: CharacterSet = {
        var s = CharacterSet.alphanumerics
        s.insert(charactersIn: "-._~")
        return s
    }()

    private static func encodeComponent(_ raw: String) -> String {
        raw.addingPercentEncoding(withAllowedCharacters: queryAllowed) ?? raw
    }
}
